File: src/main/java/leetCode/offer/Offer32_3.java
package leetCode.offer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author jingxinwu
 * @date 2021-11-21 9:07 下午
 */
public class Offer32_3 {

    public List<List<Integer>> levelOrder(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();
        if (root != null) {
            queue.add(root);
        }
        boolean flag = false;
        while (!queue.isEmpty()) {
            List<Integer> tmp = new ArrayList<>();
            for (int i = queue.size(); i > 0; i--) {
                TreeNode node = queue.poll();
                tmp.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }

            Collections.reverse(tmp);
            res.add(tmp);
            flag = !flag;
        }
        return res;
    }
}

File: src/main/GO/leetcode/lists/offer/Offer32_3.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 19:02
 */

func levelOrder32_3(root *TreeNode) [][]int {

	if root == nil {
		return [][]int{}
	}

	q := make([]*TreeNode, 0)
	q = append(q, root)
	list := make([][]int, 0)
	flag := false
	for len(q) > 0 {
		size := len(q)
		level := make([]int, 0)
		for size > 0 {
			node := q[0]
			q = q[1:]
			level = append(level, node.Val)
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
			size--
		}
		if flag {
			Reverse(level, 0, len(level)-1)
		}
		list = append(list, level)
		flag = !flag
	}
	return list
}

func Reverse(s []int, start, end int) []int {
	for i, j := start, end; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
	return s
}
